api/v1alpha1/bmcjob_types.go

```diff
@@ -0,0 +1,126 @@
+/*
+Copyright 2022 Tinkerbell.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// BMCJobConditionType represents the condition of the BMC Job.
+type BMCJobConditionType string
+
+const (
+	// JobCompleted represents successful completion of the BMC Job tasks.
+	JobCompleted BMCJobConditionType = "Completed"
+	// JobFailed represents failure in BMC job execution.
+	JobFailed BMCJobConditionType = "Failed"
+	// JobTaskCompleted represents successful completion of a BMC Job task.
+	JobTaskCompleted BMCJobConditionType = "TaskCompleted"
+)
+
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// BMCJobSpec defines the desired state of BMCJob
+type BMCJobSpec struct {
+	// Tasks represents a list of baseboard management actions to be executed.
+	// The tasks are completed in order.
+	Tasks []Task `json:"tasks"`
+}
+
+// Task represents the action to be performed.
+// A single task can only perform one type of action.
+// For example either PowerAction or BootAction.
+// +kubebuilder:validation:MaxProperties:=1
+type Task struct {
+	// PowerAction represents a baseboard management power operation.
+	PowerAction *PowerAction `json:"powerAction,omitempty"`
+
+	// BootAction represents a baseboard management boot device operation.
+	BootAction *BootAction `json:"bootAction,omitempty"`
+}
+
+type PowerAction struct {
+	// State represents the requested power state to set for the baseboard management.
+	// +kubebuilder:validation:Enum=On;Off;Status;Cycle;Reset;Soft
+	State PowerState `json:"state"`
+}
+
+type BootAction struct {
+	// BootDevice represents the requested boot device to set.
+	// +kubebuilder:validation:Enum=PXE;Disk;BIOS;CDROM;Safe
+	BootDevice BootDevice `json:"bootDevice"`
+
+	// Persistent if True, boot device is permanently set for the baseboard management.
+	// If False, boot device is set as one time boot.
+	// +kubebuilder:default=false
+	Persistent bool `json:"persistent,omitempty"`
```

Comment:
I think we probably want to represent persistent state on the `BaseboardManagement` type, which would make this always ephemeral. In that case can we remove this flag?

Reply:
I believe right now we wanted to hold off on modeling bootdevice from `BaseboardManagement` type? Since we were not sure if it would be a slice and since bmclib does not fetch boot device/order.

My intention here was to enable the job to be able to set persistent and one time bootdevice. 
Or is the suggestion to only set one time boot from job by default?

Reply:
I think this comes back to the consumers desire. A BMC controller could, for example, create an action that performs a persistent boot set or a temporary boot set using this option. 

That said, I think we can only support one time boot and introducing this later wouldn't be a breaking API change so I suggest removing it in favor of this action only ever doing a temporary boot thing then we can work out the details of configuring persisted boot order later.

Reply:
Right. So if i understood correctly, we want to default to setting one time boot through the controller.
And later extend for persistent? Maybe even delegate that to a boot order setting action.